dbms/src/Storages/Page/V3/PageDirectory.cpp

```diff
@@ -117,6 +119,88 @@ void VersionedPageEntries::createNewEntry(const PageVersion & ver, const PageEnt
         ErrorCodes::PS_DIR_APPLY_INVALID_STATUS);
 }
 
+PageIdV3Internal VersionedPageEntries::createUpsertEntry(const PageVersion & ver, const PageEntryV3 & entry)
+{
+    auto page_lock = acquireLock();
+
+    if (type == EditRecordType::VAR_DELETE)
+    {
+        type = EditRecordType::VAR_ENTRY;
+        assert(entries.empty());
+        entries.emplace(ver, EntryOrDelete::newNormalEntry(entry));
+        return buildV3Id(TEST_NAMESPACE_ID, INVALID_PAGE_ID);
+    }
+
+    if (type == EditRecordType::VAR_ENTRY)
+    {
+        auto last_iter = MapUtils::findLess(entries, PageVersion(ver.sequence + 1, 0));
+        if (last_iter == entries.end())
+        {
+            entries.emplace(ver, EntryOrDelete::newNormalEntry(entry));
+        }
+        else if (last_iter->second.isDelete())
+        {
+            // append after delete
+            entries.emplace(ver, EntryOrDelete::newNormalEntry(entry));
+        }
+        else
+        {
+            assert(last_iter->second.isEntry());
+            // It is ok to replace the entry with same sequence and newer epoch, but not valid
+            // to replace the entry with newer sequence.
+            if (unlikely(last_iter->second.being_ref_count != 1 && last_iter->first.sequence < ver.sequence))
+            {
+                throw Exception(
+                    fmt::format("Try to replace normal entry with an newer seq [ver={}] [prev_ver={}] [last_entry={}]",
+                                ver,
+                                last_iter->first,
+                                last_iter->second.toDebugString()),
+                    ErrorCodes::LOGICAL_ERROR);
+            }
+            // create a new version that inherit the `being_ref_count` of the last entry
+            entries.emplace(ver, EntryOrDelete::newReplacingEntry(last_iter->second, entry));
+        }
+        return buildV3Id(TEST_NAMESPACE_ID, INVALID_PAGE_ID);
+    }
+
+    if (type == EditRecordType::VAR_REF)
+    {
+        // an ref-page is rewritten into a normal page
+        if (!is_deleted)
+        {
+            // Full gc commit, we need to rewrite this page to
+            // be normal page with upsert-entry.
+            entries.emplace(ver, EntryOrDelete::newNormalEntry(entry));
```

Comment:
I wonder why we don't need to rewrite the real page data to BlobFile.

Reply:
We DO rewrite the real page data in `BlobStore::gc`, this function is to commit the upsert-entry in memory after we rewrite the data on disk.

1. `page_directory->getEntriesByBlobIds` we collect tuples `<ref-id, create-ver, old-entry>`
2. `blob_store.gc` rewrite the real page data to a new BlobFile, returning with tuples like `<ref-id, create-ver, upsert-entry>`
3. `page_directory->gcApply` will call this function to commit the upsert-entry in memory, turning a RefPage into Normal Page with the upsert-entry.

Reply:
Maybe this comment is better?
```
            // Full GC has rewritten data on disk, we need to update this RefPage
            // to be a normal page with the upsert-entry.
```

Reply:
I get it.